fairscale/optim/oss.py

```diff
@@ -175,25 +197,63 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
             loss = self.optim.step(**kwargs)
 
         # Sync all the states. Broadcast requests are issued async, we check completeness before moving on
-        requests = []
-        requires_grad = []
-        for rank, param_groups in enumerate(self.partition_parameters()):
-            for param_group in param_groups:
-                for param in param_group["params"]:
-                    # NOTE: Broadcast is in-place and not differentiable
-                    # Gloo will rightly assert on this operation for any tensor that requires grad.
-                    # We save and restore the grad requirement state to work around that, in our case
-                    # the grad is only useful on the source rank.
-                    global_rank = self.get_global_rank(self.group, rank)
+        for device_params in self.per_device_params:  # all the params on this device (inc all ranks)
+            self._broadcast_params_task(
+                self._buffers[device_params[0][0].device], device_params, self.group, self.global_rank
+            )
+
+        return loss
+
+    @staticmethod
+    def _broadcast_params_task(buffer: torch.Tensor, params: List[List[Parameter]], group: Any, self_rank: int) -> None:
+        """Helper function to broadcast all the parameters
+        """
+        buffer_size = buffer.numel()
+
+        for rank, rank_params in enumerate(params):  # all the params sorted per rank
+            restore_require_grad = []
 
-                    requires_grad.append((param, param.requires_grad))
+            if len(rank_params) == 0:
+                continue
+
+            global_rank = OSS.get_global_rank(group, rank)
+
+            # Copy small gradients into per-GPU buffers and then async broadcast
+            i_p = 0
+            offset = 0
+            while i_p < len(rank_params) and offset + rank_params[i_p].numel() < buffer_size:
```

Comment:
maybe I am not understanding the code right, but I thought you need an outer loop here to broadcast using the buffer potentially multiple times. I.e.

- say your buffer size is 10
- your param sizes are [2,3,3,4,4,5, 6]
- you need to pack 2,3,3 in the buffer once and then 4,4 again using the buffer and then potentially skip the single param buffer cases, etc.

I think it would be nice to refactor the code that does bucket send in a separate function and unit test it for different corner cases. What do you think?

Reply:
it's not that practical to loop over the same buffer, because it means that it cannot be async anymore (or you need to allocate many more buffers), to me it's the key issue with this approach, ditching async or needed more buffers are big drawbacks. 

So yes the proposal is to change the logic, it buckets the smallest items (we can increase the buffer size if we want to cover more of them) and sends it async, and the rest is sent as-is (not bucketed), async also except for the last one which is a gatekeeper. 

Empirically it's significantly faster

Reply:
(second point) yes for the factoring of this bucketing function, you're right, would be more readable and easier to test. Doing that

Reply:
Ah, got it. Can you please document it if not already? I feel that this will be very model dependent. If the param list has a lot of small tensors, then it can be put some in a single buffer, but the rest will be sent one by one again.

With async, does the comm library merge small messages? If so, perhaps async means we don't really need any buffer? Maybe keep it simple for now and add/decide buffering strategy later? But let me know if you want to keep as it. I can approve this version if so.

Reply:
> Ah, got it. Can you please document it if not already? I feel that this will be very model dependent. If the param list has a lot of small tensors, then it can be put some in a single buffer, but the rest will be sent one by one again.

I think that if the model has many small params, we're toasted anyway, bucketing or not. We can kind of speed up the coms if we bucket, but that requires many buffers to keep requests mostly async, and then the whole zero point (memory saving) is not that valuable. In any case bucketing over and over is not a silver bullet I think

> With async, does the comm library merge small messages? If so, perhaps async means we don't really need any buffer? Maybe keep it simple for now and add/decide buffering strategy later? But let me know if you want to keep as it. I can approve this version if so.

NCCL does not seem to bucket by itself, traces show some of the broadcast calls in the 80us range 
![Screen Shot 2020-10-01 at 4 04 34 PM](https://user-images.githubusercontent.com/3072861/94871870-e6e14980-03ff-11eb-82a8-70530558fde0.png)
A motivation for this is that currently OSS is slow when many nodes are involved, around 30% throughput hit for computer vision workloads, and the broadcast step is a big culprit (purple bar in the attached trace). We can test for real whether these changes go in the right direction of course, the benchmarks on CircleCI are not enough